Sources/MongoSwift/BSON/BSON.swift

```diff
@@ -0,0 +1,413 @@
+        guard case let .objectId(o) = self else {
+            return nil
+        }
+        return o
+    }
+
+    /// If this `BSON` is a `.date`, return it as a `Date`. Otherwise, return nil.
+    public var dateValue: Date? {
+        guard case let .date(d) = self else {
+            return nil
+        }
+        return d
+    }
+
+    /// If this `BSON` is an `.array`, return it as an `[BSON]`. Otherwise, return nil.
+    public var arrayValue: [BSON]? {
+        guard case let .array(a) = self else {
+            return nil
+        }
+        return a
+    }
+
+    /// If this `BSON` is a `.string`, return it as a `String`. Otherwise, return nil.
+    public var stringValue: String? {
+        guard case let .string(s) = self else {
+            return nil
+        }
+        return s
+    }
+
+    /// If this `BSON` is a `.document`, return it as a `Document`. Otherwise, return nil.
+    public var documentValue: Document? {
+        guard case let .document(d) = self else {
+            return nil
+        }
+        return d
+    }
+
+    /// If this `BSON` is a `.bool`, return it as an `Bool`. Otherwise, return nil.
+    public var boolValue: Bool? {
+        guard case let .bool(b) = self else {
+            return nil
+        }
+        return b
+    }
+
+    /// If this `BSON` is a `.binary`, return it as a `Binary`. Otherwise, return nil.
+    public var binaryValue: Binary? {
+        guard case let .binary(b) = self else {
+            return nil
+        }
+        return b
+    }
+
+    /// If this `BSON` is a `.double`, return it as a `Double`. Otherwise, return nil.
+    public var doubleValue: Double? {
+        guard case let .double(d) = self else {
+            return nil
+        }
+        return d
+    }
+
+    /// If this `BSON` is a `.decimal128`, return it as a `Decimal128`. Otherwise, return nil.
+    public var decimal128Value: Decimal128? {
+        guard case let .decimal128(d) = self else {
+            return nil
+        }
+        return d
+    }
+
+    /// If this `BSON` is a `.symbol`, return it as a `Symbol`. Otherwise, return nil.
+    public var symbolValue: Symbol? {
+        guard case let .symbol(s) = self else {
+            return nil
+        }
+        return s
+    }
+
+    /// If this `BSON` is a `.dbPointer`, return it as a `DBPointer`. Otherwise, return nil.
+    public var dbPointerValue: DBPointer? {
+        guard case let .dbPointer(d) = self else {
+            return nil
+        }
+        return d
+    }
+
+    /// If this `BSON` is a `.timestamp`, return it as a `Timestamp`. Otherwise, return nil.
+    public var timestampValue: Timestamp? {
+        guard case let .timestamp(t) = self else {
+            return nil
+        }
+        return t
+    }
+
+    /// Return this BSON as an `Int` if possible.
+    /// This will coerce non-integer numeric cases (e.g. `.double`) into an `Int` if such coercion would be lossless.
+    public func asInt() -> Int? {
```

Comment:
these are the helpers that coerce different `BSON` cases into Swift types. For now, I've just added ones for the numeric types and `.array`, but it can always be expanded. For instance, maybe an `asString()` could be useful for coercing `.string`, `.symbol`, and even `.objectId`.